Use `statement_descriptor_suffix` for card payments and fallback to Stripe account statement descriptions for all payments

james-allan opened the pull request:
Fixes #2830 

## Changes proposed in this Pull Request:

According to [Stripe's docs](https://support.stripe.com/questions/use-of-the-statement-descriptor-parameter-on-paymentintents-for-card-charges), using the `statement_descriptor` parameter for card payments is deprecated since January 2nd 2024. We're suppose to be using `statement_descriptor_suffix` instead. 

No one internally has been able to replicate this issue, however many merchants have already reported this and experiencing this error. 

This PR does a number of things to fix this:

1. Customisable Statement descriptors from within the plugin settings are essentially removed.
   - The Stripe plugin settings screen now pulls the statement descriptions from the Stripe Account. 
   - These fields are no longer editable. 

<p align="center">
<img width="700" alt="Screenshot 2024-01-16 at 5 35 41 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/28d3fa30-4b27-4aa4-8f57-f78b5cc4a728">
</p>

2. To make sure changes that users make to their Stripe Account are instantly reflected in the WC Settings. I've made sure that the account cache is cleared on loading the Stripe settings page. 
3. All transactions now don't pass the `statement_descriptor` arg. 
   - Stripe already sets the transaction statement description to the one set on the account anyway. 
4. If the store has enabled dynamic statement descriptions (the "Add customer order number to the bank statement" setting), Card transactions now provide that via the `statement_descriptor_suffix`.

## Testing instructions

**Base tests**

1. Go to the Stripe plugin settings. 
   - On `develop` you'll notice the bank statement description settings are customisable. 
   - On this branch those text fields are now read only and cannot be changed. I've also added some information about how merchants can update those bank statement descriptors. 

| Before | After |
|--------|--------|
| <img width="714" alt="Screenshot 2024-01-16 at 5 47 11 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/0c5ffc7e-ac3b-4c03-8f43-4d2d535a908c"> | <img width="717" alt="Screenshot 2024-01-17 at 3 57 20 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/52a6147d-372b-4a4b-83ab-9ceaa6ba349f"> | 

2. On this branch switching between checking the "Add customer order number to the bank statement" setting should still toggle the example statement descriptions.
3. Change the statement descriptions in your Stripe settings (https://dashboard.stripe.com/settings/public). Refresh the Stripe plugin settings page and notice the changes are reflected in the settings immediately. 

**Repeat the following steps with both UPE enabled and disabled.**

1. Turn **off** the "Add customer order number to the bank statement" setting.
3. Make a purchase using the Card payment method. 
4. View the transaction in your Stripe dashboard. 
    - It should include a statement descriptor which matches the setting in the Stripe settings and Stripe plugin settings
   
<p align="center">
<img width="300" alt="Screenshot 2024-01-16 at 5 53 51 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/dc26fc67-bc04-4a8a-8181-90b1bee1ddeb">
</p>

5. Place an order and use another PM like **Giropay**. 
   - Note: other APMs don't include bank statement descriptions.
6. Place an order and use **Google** or **Apple pay**.  
7. View the transaction in your Stripe dashboard. 
    - It should include a statement descriptor which matches the setting in the Stripe settings and Stripe plugin settings.
2. Turn **on** the "Add customer order number to the bank statement" setting.
3. Repeat all the example purchases I mentioned above. 
    - APM (Giropay) - Still no statement description.
    - Card payments and express payment methods (Google or Apple Pay) - you should see the statement description now includes the order number.

<p align="center">
<img width="300" alt="Screenshot 2024-01-16 at 6 02 09 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/0b76ad97-fbf8-45f5-8f8f-30f32da10079">
</p>


---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

james-allan committed:
Update Stripe statement description settings to include the date from the Stripe account

james-allan committed:
Clear account cache on the settings page load

james-allan committed:
Update the UPE gateway to make sure the statement descriptor uses the suffix for card payments

james-allan committed:
Update the non-upe gateway to make sure the statement descriptor uses the suffix for card payments

james-allan committed:
Add changelog entries

james-allan commented on includes/class-wc-stripe-helper.php:
```diff
@@ -560,6 +560,31 @@ public static function get_dynamic_statement_descriptor( $statement_descriptor =
+
+			// Stripe requires at least 1 latin (alphabet) character in the suffix so we add the first character of the prefix before the order number.
+			if ( 0 === preg_match( '/[a-zA-Z]/', $suffix ) ) {
+				$suffix = substr( $prefix, 0, 1 ) . ' ' . $suffix;
```
I just wanted to add some additional details for the reviewer. Stripe requires the `statement_descriptor_suffix` to have at least 1 latin character in it - ie an letter from the alphabet. 

Normally the `get_order_number()` wouldn't include a letter as it would just be `#1245`.  Some folks do customise the order number to include a letter so I've made this conditional based on that. 

If the order number doesn't include a letter, I just took the first letter of the prefix from the Stripe account settings and added it.

eg the suffix may look like this `WCDEV* W #123`.

I considered an "O" for order. I considered the customers first name initial. I didn't like either of those two options though. Feedback is very welcome :)

james-allan committed:
Fix js tests

james-allan committed:
Update js test to supply the statement descriptors via the account data

james-allan committed:
Update php unit tests to no longer expect a statement descriptor

james-allan committed:
Add missing "in"

james-allan commented:
I managed to actually get this error today.

<img width="1133" alt="Screenshot 2024-01-19 at 10 14 54 am" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/a4536f40-197b-41b1-a81b-efd201fcb096">

I'm not sure why but to replicate I had to use a Brazilian Stripe account. It might not be limited to Brazil merchants. Perhaps its just new users. 🤷‍♂️

james-allan commented:
@diegocurbelo I'm just noting that because `develop` has now progressed past `trunk` and I suspect we'll want to release a patch for this, I've rebased this PR and pointed it at `trunk` rather than `develop`. 

I'm still not sure what causes it. But now that I have been able replicate on own account I have a feeling it's either new accounts or specific to certain countries.

diegocurbelo reviewed (changes requested):
Thanks for all the work investigating this @james-allan!

The code looks good, and it works great (the fields updating with the Stripe on refresh are great).

The required latin character in the suffix is not ideal, and I don't have a different idea 🤷🏼, but we should update the preview in the settings to reflex that extra character. 

<img width="796" alt="Screenshot 2024-01-19 at 18 57 11" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/e5313f6f-a1cb-4514-8a77-ffb26dd56b8e">

james-allan committed:
Add latin character suffix to statement descriptor preview

james-allan committed:
Update js test to expect the suffix latin character insert

james-allan commented:
> we should update the preview in the settings to reflex that extra character.

Thanks for the feedback @diegocurbelo. I've updated the statement previews to include the latin character inserted into the description. I changed the description in my Stripe settings to confirm it is updated in the previews too. 

| <img width="728" alt="Screenshot 2024-01-22 at 6 11 12 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/83391db2-eeac-4676-b57f-8e221b32f478"> | <img width="721" alt="Screenshot 2024-01-22 at 6 03 40 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/17b5b97a-ca4b-498d-8960-38721cf3deb2"> |
|--------|--------|

diegocurbelo reviewed (approved):
Thanks for the changes @james-allan! 🚢
